### PR TITLE
Port to python3

### PR DESCRIPTION
**Pull request by sourabhaa**
This PR completes the open PR #3 and ports to python3.

@chimosky @srevinsaju @JuiP @Saumya-Mishra9129 kindly review.

### CODE REVIEW

File: math_hurdler.py [hunk 1]
```diff
@@ -148,7 +149,7 @@ def run(self):
             Color.GREEN,
             (0, 0),
             (ground.get_width(), 0),
-            ground.get_height() / 2
+            int(ground.get_height() / 2)
```
Comment (chimosky) [inline math_hurdler.py:170]: It'll also be nice to find out if the division here should be done with //, same as all similar occurrences that follow.

File: math_hurdlerActivity.py [hunk 2]
```diff
@@ -90,4 +90,5 @@ def read_file(self, file_path):
         self.game.read_file(file_path)
 
     def write_file(self, file_path):
-        self.game.write_file(file_path)
+        # self.game.write_file(file_path)
```
Comment (chimosky) [inline math_hurdlerActivity.py]: I'm guessing you forgot about this.

File: sugargame/event.py [hunk 3]
```diff
@@ -170,10 +170,10 @@ def _keyevent(self, widget, event, type):
             self.__keystate[keycode] = type == pygame.KEYDOWN
             if type == pygame.KEYUP:
                 mod = self._keymods()
-            ukey = unichr(Gdk.keyval_to_unicode(event.keyval))
+            ukey = chr(Gdk.keyval_to_unicode(event.keyval))
             if ukey == '\000':
                 ukey = ''
-            evt = pygame.event.Event(type, key=keycode, unicode=ukey, mod=mod)
+            evt = pygame.event.Event(type, key=keycode, str=ukey, mod=mod)
```
Comment (chimosky) [inline sugargame/event.py:176]: This should throw an error as [`pygame.event.Event`](https://www.pygame.org/docs/ref/event.html#pygame.event.Event) only takes two arguments, you can test this in your interpreter by calling the function with the needed arguments.

**Review by chimosky — CHANGES_REQUESTED** [review]

File: math_hurdler.py (hunk 1 above)
Comment (sourabhaa) [reply math_hurdler.py:170]: Sure. Since I'm converting float values to int, I'm anyway getting rid of vales after decimal. Flooring the values using `//` wouldn't make a difference in the result now. If I floor the values, then I think I don't need to typecast to int. Both ways give the same result, I'm not sure which is the most efficient one.

File: math_hurdlerActivity.py (hunk 2 above)
Comment (sourabhaa) [reply math_hurdlerActivity.py]: As of now, the game does not resume, it restarts. I'll get it right soon.

File: math_hurdler.py (hunk 1 above)
Comment (srevinsaju) [reply math_hurdler.py:170]: > If I floor the values, then I think I don't need to typecast to int. Both ways give the same result, I'm not sure which is the most efficient one.

Well, not exactly.

```python
>>> 3.0 // 2
1.0
>>> int(3.0 / 2)
1
```

If you need the `type` of the result to be of `int`, then, `int` is the best usage. If you don't care of the type of the output, i.e, can be either `float` or `int`, then, I would suggest using `//` instead.

File: math_hurdlerActivity.py [hunk 4]
```diff
@@ -31,15 +31,16 @@ def __init__(self, handle):
         self.build_toolbar()
 
         # Build the Pygame canvas.
-        self._pygamecanvas = sugargame.canvas.PygameCanvas(self)
+        self._pygamecanvas = sugargame.canvas.PygameCanvas(self,main = self.game.run)
```
Comment (srevinsaju) [inline math_hurdlerActivity.py]: flake.

File: math_hurdlerActivity.py [hunk 5]
```diff
@@ -31,15 +31,16 @@ def __init__(self, handle):
...
 
         # Note that set_canvas implicitly calls read_file when
         # resuming from the Journal.
         self.set_canvas(self._pygamecanvas)
 
         # Start the game running (self.game.run is called when the
         # activity constructor returns).
-        self._pygamecanvas.run_pygame(self.game.run)
+        #Deprecated:Not required in sugargame v0.12
+        #self._pygamecanvas.run_pygame(self.game.run)
```
Comment (srevinsaju) [inline math_hurdlerActivity.py]: flake. `#` should follow a whitespace character

File: sugargame/__init__.py [hunk 6]
```diff
@@ -1 +1 @@
-__version__ = '1.1'
+__version__ = '1.2'
```
Comment (srevinsaju) [inline sugargame/__init__.py]: Btw, why aren't we porting this activity to Sugargame 1.3 again? I must be missing something

**Review by srevinsaju — CHANGES_REQUESTED** [review]
Minor changes. Looks good overall.

File: math_hurdler.py (hunk 1 above)
Comment (sourabhaa) [reply math_hurdler.py:170]: okay, we need `int` type here. In this case, would I still need to floor it?

File: sugargame/__init__.py (hunk 6 above)
Comment (sourabhaa) [reply sugargame/__init__.py]: Yes we need to, I will do it now!

File: math_hurdler.py (hunk 1 above)
Comment (chimosky) [reply math_hurdler.py:170]: @sourabhaa said;
> okay, we need int type here. In this case, would I still need to floor it?

No you don't need to use floor division in that case.

File: math_hurdlerActivity.py (hunk 2 above)
Comment (chimosky) [reply math_hurdlerActivity.py]: When you say it restarts, I'm guessing you mean you're yet to implement saving to journal correctly so when the activity instance resumes no values have been stored to be restored so instead you just start a new instance, is this correct?

File: bezierHelper.py [hunk 7]
```diff
@@ -1,82 +1,90 @@
-import os, sys
+import os
+import sys
```
Comment (chimosky) [inline bezierHelper.py:2]: Is there any particular reason for changing this?

File: bezierHelper.py [hunk 8]
```diff
@@ -1,82 +1,90 @@
...
         self.v3 = v3
-        self.t = self.Clamp01(t/duration)
+        self.t = self.Clamp01(t / duration)
         oneMinusT = 1 - self.t
-        return (oneMinusT * oneMinusT * v1 + \
-        2 * oneMinusT * self.t * v2 + \
-        self.t * self.t * v3)
+        return (oneMinusT * oneMinusT * v1 +
+                2 * oneMinusT * self.t * v2 +
+                self.t * self.t * v3)
```
Comment (chimosky) [inline bezierHelper.py:32]: pep8 says the operator should be on the same line as the next operand so in this case `+ 2` is preferable to having the operator on the previous line and it also helps with readability.

File: bezierHelper.py [hunk 9]
```diff
@@ -1,82 +1,90 @@
...
+        self.t = self.Clamp01(t / duration)
         oneMinusT = 1 - self.t
-        return (oneMinusT * oneMinusT * oneMinusT * v1 + \
-        3 * oneMinusT * oneMinusT * self.t * v2 + \
-        3 * oneMinusT * self.t * self.t * v3 + \
-        self.t * self.t * self.t * v4)
+        return (oneMinusT * oneMinusT * oneMinusT * v1 +
+                3 * oneMinusT * oneMinusT * self.t * v2 +
+                3 * oneMinusT * self.t * self.t * v3 +
+                self.t * self.t * self.t * v4)
```
Comment (chimosky) [inline bezierHelper.py:46]: Same comment as above.

File: math_hurdler.py [hunk 10]
```diff
@@ -1,18 +1,18 @@
...
 import gi
-gi.require_version('Gtk','3.0')
-from gi.repository import Gtk
+gi.require_version('Gtk', '3.0')
 
-from sprites.sun import Sun
-from sprites.horse import Horse
-from objects.button import Button
-from question import Question
-from fractions import Fraction
```
Comment (chimosky) [inline math_hurdler.py]: Why change the order of imports?

File: math_hurdler.py [hunk 11]
```diff
@@ -155,18 +157,23 @@ def run(self):
...
         sun = Sun()
-        sun.rect.topleft = (screen_size[0]- sun.image.get_width(),0)
+        sun.rect.topleft = (screen_size[0] - sun.image.get_width(), 0)
 
         horse = Horse()
-        horse.rect.x = display_info.current_h/3
-        horse.rect.y = display_info.current_h - horse.image.get_height() - ground.get_height()
+        horse.rect.x = display_info.current_h / 3
+        horse.rect.y = display_info.current_h - \
+            horse.image.get_height() - ground.get_height()
```
Comment (chimosky) [inline math_hurdler.py:181]: Same comment as above.

File: math_hurdler.py [hunk 12]
```diff
@@ -155,18 +157,23 @@ def run(self):
...
 
         hurdle = pygame.image.load('./assets/images/hurdle.png')
-        hurdle = pygame.transform.scale(hurdle,(int(hurdle.get_height()/3),int(hurdle.get_width()/3)))
+        hurdle = pygame.transform.scale(
+            hurdle,
+            (int(hurdle.get_height() / 3), int(hurdle.get_width() / 3)))
 
-        hurdle_y = display_info.current_h - hurdle.get_height() - (2*ground.get_height()/3)
+        hurdle_y = display_info.current_h - \
+            hurdle.get_height() - (2 * ground.get_height() / 3)
```
Comment (chimosky) [inline math_hurdler.py:189]: This too.

File: math_hurdler.py [hunk 13]
```diff
@@ -298,64 +314,79 @@ def evaluate_answer(answer):
                     self.horse_change += 1
 
                     horse.rect.x = display_info.current_w / 3
-                    horse.rect.y = display_info.current_h - hurdle.get_height() - (3*ground.get_height()/4)
+                    horse.rect.y = display_info.current_h - \
+                        hurdle.get_height() - (3 * ground.get_height() / 4)
```
Comment (chimosky) [inline math_hurdler.py:335]: Same comment as the ones before it.

File: math_hurdler.py [hunk 14]
```diff
@@ -298,64 +314,79 @@ def evaluate_answer(answer):
...
+                question_board.blit(
+                    self.question_text_label,
+                    (10, self.question_label.get_height() + 10))
 
                 screen.blit(ground, (0, screen_size[1] - ground.get_height()))
-                button_panel_x = ground.get_width()/4
-                button_panel_y = screen_size[1] - ground.get_height() + ground.get_height() / 3 + 10
+                button_panel_x = ground.get_width() / 4
+                button_panel_y = screen_size[1] - \
+                    ground.get_height() + ground.get_height() / 3 + 10
```
Comment (chimosky) [inline math_hurdler.py:381]: This too, and all other occurrences.

File: math_hurdlerActivity.py [hunk 15]
```diff
@@ -1,21 +1,17 @@
...
-import sugar3.activity.activity
-from sugar3.graphics.toolbarbox import ToolbarBox
-from sugar3.activity.widgets import ActivityToolbarButton
-from sugar3.graphics.toolbutton import ToolButton
-from sugar3.activity.widgets import StopButton
-
-
-import sugargame.canvas
-
-import math_hurdler
```
Comment (chimosky) [inline math_hurdlerActivity.py]: Why the change of imports too?

File: question.py [hunk 16]
```diff
@@ -1,57 +1,58 @@
...
+
+        for num in range(0, 4):
             if num == self.answer_index:
                 self.left_side = self.get_fraction()
                 self.right_side = self.get_fraction()
                 self.answer = self.left_side + self.right_side
                 while (
-                    any([x == self.answer for x in self.choices]) 
+                    any([x == self.answer for x in self.choices])
                 ):
```
Comment (chimosky) [inline question.py]: The indentation here is off and the parentheses covering the any isn't needed too, the indentation is off as the any is same indent as the body of the loop so it's confusing.

File: question.py [hunk 17]
```diff
@@ -1,57 +1,58 @@
...
                 )
-                
+
             else:
                 choice = self.create_choice()
                 while (
-                    any([x == choice for x in self.choices]) 
+                    any([x == choice for x in self.choices])
                 ):
                     choice = self.create_choice()
```
Comment (chimosky) [inline question.py]: This too.

**Review by chimosky — CHANGES_REQUESTED** [review]

**Comment by chimosky** [comment]
Reviewed 6551aa1.

**Comment by sourabhaa** [comment]
I ran `autopep8` against all the files. Hence, all these changes. I didn't inspect manually except for fixing E501 (line too long) errors. Shouldn't I be doing this way?

File: math_hurdlerActivity.py (hunk 2 above)
Comment (sourabhaa) [reply math_hurdlerActivity.py]: Yes, I did it temporarily until that feature is ready.

File: sugargame/event.py (hunk 3 above)
Comment (sourabhaa) [reply sugargame/event.py:176]: I went through the documentation and also I checked many activities. This exists in all activities wrapped with `sugargame` and it seems to work fine. I'm not sure if there is a need to alter it. Also, I'm not fully aware of every argument in that line, So I don't know how make any changes to get it right (if required).

**Comment by chimosky** [comment]
> I ran `autopep8` against all the files. Hence, all these changes. I didn't inspect manually except for fixing E501 (line too long) errors. Shouldn't I be doing this way?

Yeah you should so you can keep existing conventions and make some other necessary changes, you can see what [pep8 says about breaking a line before or after a binary operator](https://pep8.org/#break-before-or-after-binary-operator).

**Comment by srevinsaju** [comment]
> I ran autopep8 against all the files. Hence, all these changes. I didn't inspect manually except for fixing E501 (line too long) errors. Shouldn't I be doing this way?

By doing this, you are also changing the existing code. This makes debugging any possible regression which is introduced by this code difficult. To be clear, what I mean by the flake errors I pointed out: is the formatting of the 'newly added code', not the 'already existing code'. 

Flake8 for the entire source code is welcome, but I would prefer to see it in another PR, because while squashing your commits, these changes go all together, and makes it hard to `git blame` in future 

TLDR
1. New changes should adhere to flake8
2. Changing the style of previous code should happen in a separate PR, not in a `port-to-python3` PR. :) 


Reviewed https://github.com/sugarlabs/math-hurdler/commit/6551aa13cc44df3465f0646b2f1b7faa336236ec

**Comment by quozl** [comment]
On the other hand, don't let the perfect be the enemy of the good.

It really depends on what your reviewers want ... and that can be separate patches for seperate changes, keep existing code style, or no code style changes at all.

Combining everything into one patch or one pull request is asking for more reviewer effort.

Sorry, I've not had the time to review this pull request.

**Comment by sourabhaa** [comment]
Thanks for clearing it @chimosky @srevinsaju @quozl. I'll make the necessary changes.

File: math_hurdlerActivity.py (hunk 2 above)
Comment (sourabhaa) [reply math_hurdlerActivity.py]: Should I add this feature in this PR or another PR?

File: sugargame/event.py (hunk 3 above)
Comment (srevinsaju) [reply sugargame/event.py:176]: > This should throw an error as pygame.event.Event only takes two arguments, you can test this in your interpreter by calling the function with the needed arguments.

`pygame.event.Event` allows two signatures
```python
Event(type, dict) -> EventType instance
Event(type, **attributes) -> EventType instance
```

I think @sourabhaa has used the second signature in this (`**kwargs`) which allows `n` number of key=value parameters to `pygame.event.Event`. imo, it shouldn't throw an error :thinking: .

**Comment by sourabhaa** [comment]
I've made flake8 fixes to the new changes. @srevinsaju @chimosky  @JuiP @Saumya-Mishra9129 Kindly review.

File: math_hurdlerActivity.py (hunk 2 above)
Comment (chimosky) [reply math_hurdlerActivity.py]: You can add it here after fixing the existing issues here.

File: sugargame/event.py (hunk 3 above)
Comment (chimosky) [reply sugargame/event.py:176]: Okay, that might be the case but I'd ran it in an interpreter and got an error. I haven't tested the activity yet to know if it throws an error.

**Comment by chimosky** [comment]
Reviewed f6c3223. Not tested.

**Comment by srevinsaju** [comment]
Reviewed https://github.com/sugarlabs/math-hurdler/pull/4/commits/f6c32237c9e4b1942eed3a214b11449b2694ee36, looks better. Good work ✨

**Comment by sourabhaa** [comment]
@chimosky @srevinsaju @JuiP @Saumya-Mishra9129 kindly review. Should I make commits related to release in this PR?

**Comment by srevinsaju** [comment]
i have previously seen GSoC students doing releases in a separate PR. @chimosky thoughts?

**Comment by chimosky** [comment]
> i have previously seen GSoC students doing releases in a separate PR. @chimosky thoughts?

Agreed, a release should be done in a different PR as this was for a port to Python 3 and that's all it should be.

File: math_hurdlerActivity.py [hunk 18]
```diff
@@ -88,8 +84,14 @@ def _stop_play_cb(self, button):
...
+            score, play_state = [restore_data[i] for i in (0, 1)]
+            self.game.restore_game(score, play_state)
+            
     def write_file(self, file_path):
-        # self.game.write_file(file_path)
-        pass
+        score, play_state = self.game.write_file()
+        with open(file_path, 'w') as f:    
+            f.write(f'{score}\n{play_state}')
+
```
Comment (chimosky) [inline math_hurdlerActivity.py]: Could you remove the no new line at end of file?

**Review by chimosky — CHANGES_REQUESTED** [review]

File: math_hurdlerActivity.py (hunk 18 above)
Comment (sourabhaa) [reply math_hurdlerActivity.py]: done

**Comment by sourabhaa** [comment]
@chimosky @JuiP @Saumya-Mishra9129 @srevinsaju kindly review. Can I move to open  release PR?

File: math_hurdlerActivity.py (hunk 2 above)
Comment (sourabhaa) [reply math_hurdlerActivity.py]: done

**Comment by chimosky** [comment]
Tested, works as expected. I like the activity, you have to be quite fast with the calculations, would be nice to increase the time before each hurdle by a couple of seconds.

This is shown in the logs;
```python3
/home/ibiam/Activities/math-hurdler/math_hurdler.py:8: PyGIWarning: Gtk was imported without specifying a version first. 
Use gi.require_version('Gtk', '3.0') before import to ensure that the right version gets loaded.
  from gi.repository import Gtk
```

Fix the `PyGIWarning` and I'll merge. Thanks.

**Comment by chimosky** [comment]
Also make a change to `setup.py` and change `python` to `python3` in the shebang line as it'll default to python2 for systems with python2 as the default interpreter, debian still has it this way.

**Comment by chimosky** [comment]
Reviewed 7f9a759, thanks.

Merged by hand.
